### PR TITLE
feat: menu-component

### PR DESCRIPTION
**Pull request by olegdolenchuk**
Screenshot 2020-04-18 at 20 12 39
Screenshot 2020-04-21 at 14 49 49

### CODE REVIEW

**Comment by oknelvapi** [comment]
What's your attitude towards moving Menu, SearchBar, and Navigation Components from the 'Header' folder to the 'components' folder? The main reason to do it is avoiding deep nesting and better code reading.

**Comment by olegdolenchuk** [comment]
_> What's your attitude towards moving Menu, SearchBar, and Navigation Components from the 'Header' folder to the 'components' folder? The main reason to do it is avoiding deep nesting and better code reading._

Look at the design, all these elements belong only to the header, if we start to move all the subcomponents to the main folder we will get more folders, this does not make sense

**Comment by oknelvapi** [comment]
> Look at the design, all these elements belong only to the header, if we start to move all the subcomponents to the main folder we will get more folders, this does not make sense

Up to you :)

File: src/shared/components/Header/Menu/component.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                    return (
+                      <li key={i}>
+                        <a href={item.link}>{item.name}</a>
+                      </li>
+                    );
+                  })}
+                </ul>
+              </nav>
+            );
+          })}
```
Comment (oknelvapi) [inline src/shared/components/Header/Menu/component.tsx:71]: If you don't have any side effects, You can return jsx without `return`:
```javascript
{MenuLists.map((MenuList: MenuList, index: number) => (
            <nav key={index}>
              <strong>{MenuList.heading}</strong>
              <ul>
                {MenuList.list.map((item: MenuItem, i: number) => (
                  <li key={i}>
                    <a href={item.link}>{item.name}</a>
                  </li>
                ))}
              </ul>
            </nav>
          ))}
`

Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/component.tsx]: Done!

File: src/shared/colors.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ enum Colors {
   BODY_TEXT = '#575757',
   BLACK = '#000000',
   WHITE = '#FFFFFF',
+  PLACEHOLDER = '#b1b1b1',
```
Comment (braindrain9) [inline src/shared/colors.ts:18]: **let's not call** COLOR variable just by the place where they used - further it's not comfortable to re-use them, - as they lose their initial meaning. (placeholder could be any color - LIGHT_GREY brings right sence).

This color is already used in master like LIGHT_GREY - please reuse it.

File: src/shared/components/Header/Menu/component.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+type MenuProps = {
+  isOpenMenu: boolean;
+};
+
+type MenuItem = {
+  name: string;
+  link: string;
+};
+
+type MenuList = {
```
Comment (braindrain9) [inline src/shared/components/Header/Menu/component.tsx:15]: let's move this MenuList and MenuItem to the shared types - and import and use it here and in footer.

File: src/shared/components/Header/Navigation/component.tsx [hunk 4]
```diff
@@ -1,15 +1,26 @@
 import React from 'react';
 
+import Button from 'shared/components/Button';
```
Comment (braindrain9) [inline src/shared/components/Header/Navigation/component.tsx:3]: let's import Button to `shared/component/index.ts` and then call it from there. like

```js
import Button from 'shared/components';
```

File: src/shared/components/Header/styles.ts [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
+type StyleProps = {
+  isOpenMenu: boolean;
+};
+
+export const HeaderWrapper = styled.header<StyleProps>`
   position: relative;
   display: flex;
   align-items: center;
   height: 80px;
+  ${({ isOpenMenu }: StyleProps): string => `background: ${isOpenMenu ? '#fff' : 'transparent'};`};
```
Comment (braindrain9) [inline src/shared/components/Header/styles.ts:13]: please do not use `#fff` --> `${Colors.WHITE}`

File: src/shared/components/Header/Menu/styles.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import Colors from 'shared/colors';
+
+type StyleProps = {
+  isOpenMenu: boolean;
+};
+
+export const MenuWrapper = styled.div<StyleProps>`
+  display: block;
+  position: absolute;
+  padding: 40px 0 70px 0;
```
Comment (braindrain9) [inline src/shared/components/Header/Menu/styles.ts:11]: could be `padding: 40px 0 70px;`

File: src/shared/components/Header/Navigation/styles.ts [hunk 7]
```diff
@@ -1,5 +1,10 @@
 import styled from 'styled-components';
 import Colors from 'shared/colors';
+
+type StyleProps = {
```
Comment (braindrain9) [inline src/shared/components/Header/Navigation/styles.ts:4]: imports should go first. please note that during merging conflicts.

File: src/shared/colors.ts (hunk 2 above)
Comment (olegdolenchuk) [reply src/shared/colors.ts]: Done!

File: src/shared/components/Header/Menu/component.tsx (hunk 3 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/component.tsx]: Done!

File: src/shared/components/Header/Navigation/component.tsx (hunk 4 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/component.tsx]: Done!

File: src/shared/components/Header/styles.ts (hunk 5 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/styles.ts]: Done!

File: src/shared/components/Header/Menu/styles.ts (hunk 6 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/styles.ts]: Done!

File: src/shared/components/Footer/FooterList/component.tsx [hunk 8]
```diff
@@ -1,16 +1,9 @@
...
-  link: string;
-};
+import { MenuItem, MenuList } from 'shared/types';
 
-type FooterList = {
-  heading: string;
-  list: FooterItem[];
-};
+import { FooterListWrapper, FooterListHeading, SocialIconWrapper, FooterListItem, SocialIconLink } from './styles';
+import { TwitterIcon, LinkedinIcon, FacebookIcon } from 'shared/icons';
```
Comment (lucypoly) [inline src/shared/components/Footer/FooterList/component.tsx:6]: I think this import could be a bit higher :)

File: src/shared/components/Header/Navigation/styles.ts [hunk 9]
```diff
@@ -55,11 +83,27 @@ export const ToggleMenu = styled.button`
     width: 100%;
     height: 2px;
     background-color: ${Colors.BLACK};
+    transition: all 0.25s ease-out;
   }
   &::after {
-    top: 6px;
+    ${({ isOpenMenu }: StyleProps): string =>
```
Comment (lucypoly) [inline src/shared/components/Header/Navigation/styles.ts:89]: May be you could try 
```
${p => `bottom: ${p.isOpenMenu ? 
```
```
${({ isOpenMenu }: StyleProps): string =>
`bottom: ${isOpenMenu ? 
```

File: src/shared/components/Header/Menu/MenuListItem/component.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+  const onToggleHandler = (): void => {
+    setIsOpenList(!isOpenList);
+  };
+
+  return (
+    <nav>
+      <strong
+        onClick={(): void => {
+          onToggleHandler();
```
Comment (lucypoly) [inline src/shared/components/Header/Menu/MenuListItem/component.tsx:23]: Wh do you need one more function, but not ` onClick={() => setIsOpenList(!isOpenList);}`?

File: src/shared/components/Header/Menu/MenuListItem/component.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        }}
+      >
+        {list.heading}
+        <ArrowIconWrapper isOpenList={isOpenList}>
+          <ArrowIcon />
+        </ArrowIconWrapper>
+      </strong>
+      <ListWrapper isOpenList={isOpenList}>
+        {list.list.map((item: MenuItem, i: number) => (
+          <li key={i}>
```
Comment (lucypoly) [inline src/shared/components/Header/Menu/MenuListItem/component.tsx:33]: Maybe we could use library for keys https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318

File: src/shared/components/Footer/FooterList/component.tsx (hunk 8 above)
Comment (olegdolenchuk) [reply src/shared/components/Footer/FooterList/component.tsx:6]: I did not understand you a little. It looks like this for me
Screenshot 2020-04-22 at 11 58 02

File: src/shared/components/Header/Navigation/styles.ts (hunk 9 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/styles.ts:89]: I have a mistake
Screenshot 2020-04-22 at 12 04 01

File: src/shared/components/Header/Menu/MenuListItem/component.tsx (hunk 11 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/MenuListItem/component.tsx:33]: Good idea, but I think that there will be no problem when we have API. but maybe I'm wrong

File: src/shared/components/Footer/FooterList/component.tsx (hunk 8 above)
Comment (lucypoly) [reply src/shared/components/Footer/FooterList/component.tsx:6]: Due to our code style, absolute imports should go higher than relative

File: src/shared/components/Header/Navigation/styles.ts (hunk 9 above)
Comment (lucypoly) [reply src/shared/components/Header/Navigation/styles.ts:89]: This should probably help you https://stackoverflow.com/questions/57852188/missing-return-type-on-function-error-in-styled-component-interpolation-functi

File: src/shared/components/Header/Menu/MenuListItem/component.tsx (hunk 11 above)
Comment (lucypoly) [reply src/shared/components/Header/Menu/MenuListItem/component.tsx:33]: Problems could begin if the order of items in the list change, that's why it's not the good practice to use `index` for `key`

File: src/shared/components/Header/Menu/MenuListItem/component.tsx (hunk 10 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/MenuListItem/component.tsx]: Done!

File: src/shared/components/Header/Navigation/styles.ts (hunk 9 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/styles.ts]: Done!

File: src/shared/components/Footer/FooterList/component.tsx (hunk 8 above)
Comment (olegdolenchuk) [reply src/shared/components/Footer/FooterList/component.tsx:3]: Donee!

File: src/shared/components/Header/Menu/MenuListItem/component.tsx (hunk 11 above)
Comment (braindrain9) [reply src/shared/components/Header/Menu/MenuListItem/component.tsx:29]: good call 👍  I think we can just add some unique `id` property to our list.
e.g.
```js
list: [
   { name: 'Як це працює', link: '/' , id: Math.random()},
...
]
```

File: src/shared/components/Header/Menu/component.tsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+      document.body.style.overflow = 'unset';
+    }
+  }, [isOpenMenu]);
+
+  return (
+    <MenuWrapper isOpenMenu={isOpenMenu}>
+      <Container>
+        <MenuListWrapper>
+          {MenuLists.map((MenuList: MenuList, index: number) => (
+            <MenuListItem list={MenuList} key={index} />
```
Comment (braindrain9) [inline src/shared/components/Header/Menu/component.tsx:59]: if we are using `id` prop, we can remove passing key={index}

File: src/shared/components/Header/Navigation/styles.ts [hunk 13]
```diff
@@ -1,6 +1,10 @@
 import styled from 'styled-components';
-import Colors from 'shared/colors';
 import media from 'shared/media-types';
+import Colors from 'shared/colors';
```
Comment (braindrain9) [inline src/shared/components/Header/Navigation/styles.ts:3]: why did this import change (from 2 to 3 row)? please avoid unnecessary changes to reduce merge conflicts

File: src/shared/icons/index.ts [hunk 14]
```diff
@@ -6,5 +6,16 @@ import LogoWhiteIcon from './LogoWhite';
 import LinkedinIcon from './Linkedin';
 import TwitterIcon from './Twitter';
 import FacebookIcon from './Facebook';
+import ArrowIcon from './Arrow';
 
-export { SearchIcon, CloseIcon, LogoBlackIcon, BoxIcon, LogoWhiteIcon, LinkedinIcon, TwitterIcon, FacebookIcon };
+export {
+  SearchIcon,
```
Comment (braindrain9) [inline src/shared/icons/index.ts:12]: let's put them in alphabetical order

File: src/shared/components/Header/Menu/component.tsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+      document.body.style.overflow = 'unset';
+    }
+  }, [isOpenMenu]);
+
+  return (
+    <MenuWrapper isOpenMenu={isOpenMenu}>
+      <Container>
+        <MenuListWrapper>
+          {MenuLists.map((MenuList: MenuList, index: number) => (
```
Comment (braindrain9) [inline src/shared/components/Header/Menu/component.tsx:58]: we do not call variable uppercased - please update
```js
{MenuLists.map((list: MenuList, index: number) => (
```

File: src/shared/components/Header/Menu/MenuListItem/component.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
+import React, { useState } from 'react';
+
+import { MenuList, MenuItem } from 'shared/types';
+import { ArrowIcon } from 'shared/icons';
+
+import { ArrowIconWrapper, ListWrapper } from './styles';
+
+type MenuListItem = {
```
Comment (braindrain9) [inline src/shared/components/Header/Menu/MenuListItem/component.tsx:8]: do we really need a new type here? can't we just use `MenuList` here?

File: src/shared/components/Header/Menu/MenuListItem/component.tsx (hunk 11 above)
Comment (braindrain9) [reply src/shared/components/Header/Menu/MenuListItem/component.tsx:29]: please update key={index} to key={item.id} for footer-list too.

File: src/shared/components/Header/Menu/component.tsx (hunk 15 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/component.tsx]: Done

File: src/shared/icons/index.ts (hunk 14 above)
Comment (olegdolenchuk) [reply src/shared/icons/index.ts]: Done

File: src/shared/components/Header/Navigation/styles.ts (hunk 13 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/styles.ts:2]: Done

File: src/shared/components/Header/Menu/component.tsx (hunk 12 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/component.tsx]: Done

File: src/shared/components/Header/Menu/MenuListItem/component.tsx (hunk 11 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Menu/MenuListItem/component.tsx]: Done

File: src/shared/components/Footer/FooterList/component.tsx [hunk 17]
```diff
@@ -30,46 +21,49 @@ const calculateFlexOrder = (index: number, arrayLength: number): number => {
...
   return (
     <FooterListWrapper>
-      {footerLists.map((footerList: FooterList, index: number) => (
-        <div key={index}>
+      {footerLists.map((footerList: MenuList, index: number) => (
+        <div key={footerList.id}>
           <FooterListHeading>{footerList.heading}</FooterListHeading>
           <ul>
-            {footerList.list.map((item: FooterItem, i: number) => (
+            {footerList.list.map((item: MenuItem, i: number) => (
```
Comment (braindrain9) [inline src/shared/components/Footer/FooterList/component.tsx:64]: `i: number` could be removed

File: src/shared/components/Header/Navigation/component.tsx [hunk 18]
```diff
@@ -1,37 +1,52 @@
...
 
+import { Button } from 'shared/components';
+import { ButtonVariant } from 'shared/components/Button/styles';
+
 import { SearchIcon } from 'shared/icons';
-import { MenuListItem } from '../types';
-import { Navigation, ToggleMenu, SearchButton } from './styles';
+import { Navigation, ToggleMenu, SearchButton, ButtonWrapper } from './styles';
+
+export type NavListItem = {
```
Comment (braindrain9) [inline src/shared/components/Header/Navigation/component.tsx:9]: could we use `MenuListItem` type here?

File: src/shared/components/Header/Navigation/styles.ts [hunk 19]
```diff
@@ -36,9 +44,29 @@ export const SearchButton = styled.button`
...
+  ${media.tabletMMax} {
+    order: 2;
+  }
+  ${media.mobileMax} {
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    width: 37px;
+    height: 37px;
+    font-size: 0;
```
Comment (braindrain9) [inline src/shared/components/Header/Navigation/styles.ts:60]: `font-size: 0;` looks a bit confusing to me? maybe to use some visibility: hidden?

File: src/shared/components/Header/component.tsx [hunk 20]
```diff
@@ -3,37 +3,40 @@ import { Container, Row, Col } from 'react-bootstrap';
...
-    { name: 'Знайти проекти', link: '#' },
-    { name: 'Додати проект', link: '/link2' },
-  ];
   const [isOpenSearch, setIsOpenSearch] = useState(false);
+  const [isOpenMenu, setIsOpenMenu] = useState(false);
   const onToggleHandler = (type: string): void => {
-    if (type === 'search') {
-      setIsOpenSearch(!isOpenSearch);
+    switch (type) {
+      case 'search':
```
Comment (braindrain9) [inline src/shared/components/Header/component.tsx:17]: let's be more consistent - please put 'search' and 'menu' into some `enum` and use it where needed.

File: src/shared/components/Header/Navigation/styles.ts (hunk 19 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/styles.ts:60]: `visibility: hidden` this will completely hide the search button

File: src/shared/components/Footer/FooterList/component.tsx (hunk 17 above)
Comment (olegdolenchuk) [reply src/shared/components/Footer/FooterList/component.tsx:64]: Isn't that necessary for your order function?

File: src/shared/components/Header/component.tsx (hunk 20 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/component.tsx]: Done!

File: src/shared/components/Footer/FooterList/component.tsx (hunk 17 above)
Comment (braindrain9) [reply src/shared/components/Footer/FooterList/component.tsx:64]: right 👍 sorry, missed that 👍

File: src/shared/components/Header/Navigation/styles.ts [hunk 21]
```diff
@@ -23,10 +24,17 @@ export const Navigation = styled.nav`
       color: ${Colors.BLACK};
       text-decoration: none;
     }
-    svg {
-      margin-right: 10px;
+  }
+  ${media.tabletLMax} {
+    ul {
+      display: none;
     }
```
Comment (braindrain9) [inline src/shared/components/Header/Navigation/styles.ts:31]: is there should be `tabletMax` or `tabletLMax`?

File: src/shared/components/Header/component.tsx [hunk 22]
```diff
@@ -3,37 +3,45 @@ import { Container, Row, Col } from 'react-bootstrap';
...
-            <LogoWrapper href="#">
-              <LogoBlackIcon mobile={true} />
-              <LogoBlackIcon mobile={false} />
-            </LogoWrapper>
-          </Col>
-          <Col xs={7} sm={9}>
-            <Navigation onToggle={onToggleHandler} menuList={menuList} />
+          <Col>
+            <HeaderElementsWrapper>
+              <LogoWrapper href="#">
```
Comment (braindrain9) [inline src/shared/components/Header/component.tsx:39]: I think LogoWrapper should link to the default page `href="/"`

**Comment by braindrain9** [comment]
please fix typo in PR name : `feat: menu-compont` --> `feat: menu-component`. also it's better to add some action was made - --> `feat: add menu-component` or `feat: update menu-component` etc.

File: src/shared/components/Header/Navigation/styles.ts (hunk 21 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/styles.ts:31]: tabletLMax

**Comment by braindrain9** [comment]
also this change hasn't been resolved https://github.com/MecenatOrg/frontend/pull/13/files#r415424209

File: src/shared/components/Header/component.tsx (hunk 22 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/component.tsx]: Done!

File: src/shared/components/Header/Navigation/component.tsx (hunk 18 above)
Comment (olegdolenchuk) [reply src/shared/components/Header/Navigation/component.tsx]: Done!
